### PR TITLE
Fixed #31233 -- Closed database connections and cursors after use.

### PR DESCRIPTION
**Pull request by jdufresne**
https://code.djangoproject.com/ticket/31233

### CODE REVIEW

**Comment by jdufresne** [comment]
These cases were discovered using a psycopg2 patch I'm working on to emit a warning on unclosed resources.

https://github.com/psycopg/psycopg2/pull/1039

**Comment by jdufresne** [comment]
A couple thoughts I had while making this change:

- Perhaps `._nodb_connection` shouldn't be a property to help signal to the caller that it has additional side-effects.

- Perhaps `._nodb_connection` should wrap its return value in `contextlib.closing()`. This would force all callers to use a context manager to close it after its use is done. The connection is always intended to be temporary and short-lived anyway.

Any thoughts on these changes as part of this PR?

**Comment by claudep** [comment]
For me, the `try, finally` syntax you introduced in replacement of the `with` syntax is rather ugly, but of course I understand why you did it. If you could find a way to replace `_nodb_connection.cursor()` by some other context manager utility that auto-closes the `_nodb_connection`, it would be really nice.

**Comment by jdufresne** [comment]
The `_nodb_connection` is only ever used to access a cursor. We could change the method to `_nodb_cursor()` and return a cursor context manager instead. Inside the `_nodb_cusror()` method, the connection resource could be handled and closed upon completion. WDYT?

**Comment by claudep** [comment]
Looks promising!

**Comment by pope1ni** [comment]
Sounds reasonable, and would remove a lot of messy try-except patterns. I was going to suggest looking into `contextlib.ExitStack`, but your suggestion is better.

**Comment by jdufresne** [comment]
I have implemented the above suggestion. Thanks for the feedback.

**Comment by claudep** [comment]
Awesome, thanks! Now if you could split the commit so as the `no_db` stuff is contained in a separate commit, would be great.

**Comment by jdufresne** [comment]
Done. I also added a blurb to the "Database backend API" section of the release notes.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/db/models/query.py [hunk 1]
```diff
@@ -747,7 +747,10 @@ def _raw_delete(self, using):
         query = self.query.clone()
         query.__class__ = sql.DeleteQuery
         cursor = query.get_compiler(using).execute_sql(CURSOR)
-        return cursor.rowcount if cursor else 0
+        if cursor:
+            with cursor:
+                return cursor.rowcount
+        return 0
```
[the remark this answers is not in the document]
Comment (jdufresne) [reply django/db/models/query.py]: Is this safe to do? `execute_sql` sometimes returns `None`:

https://github.com/django/django/blob/2d55cb5c4a9c736d04290c9f99ebcbe9e196ac97/django/db/models/sql/compiler.py#L1127

[the remark this answers is not in the document]
Comment (felixxm) [reply django/db/models/query.py]: It cannot return `None` for `DeleteQuery` but I agree that custom compilers in 3rd-party  database backends may behave differently. I will revert these changes for compilers.

**Comment by jdufresne** [comment]
Thanks for finding the additional cases!

**Pull request merged**